themes/bootstrap3/js/facets.js

```diff
@@ -116,18 +116,274 @@ VuFind.register('facetList', function FacetList() {
+    for (let dateSelectorId of dateSelectorIds) {
+      dateParams = [];
+      allEmptyDateParams = true;
+      form = document.querySelector('form#' + dateSelectorId);
+      dfinputs = form.querySelectorAll('.date-fields input');
+      for (let input of dfinputs) {
+        if (window.location.search.match(input.name)) {
+          // If the parameter is already present we update it
+          updateInitialParams(input.name, input.value);
+        } else {
+          dateParams.push(encodeURI(input.name + '=' + input.value));
+        }
+        if (input.value !== '') {
+          allEmptyDateParams = false;
+        }
+      }
+
+      // If at least one parameter is not null we continue the routine for the final URL
+      if (allEmptyDateParams === false) {
+        globalAddedParams = globalAddedParams.concat(dateParams);
+        inputs = form.querySelectorAll('input');
+        for (let input of inputs) {
+          if (dateSelectorId.startsWith(input.value)) {
+            let dateRangeParam = encodeURI(input.name + '=' + input.value);
+            if (!window.location.search.match(dateRangeParam)) {
+              globalAddedParams.push(dateRangeParam);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  function getHrefWithNewParams() {
+    handleDateSelector();
+    // Unique parameters
+    initialRawParams = initialRawParams.filter(function onlyUnique(value, index, array) {
+      return array.indexOf(value) === index;
+    });
+    // Removing parameters
+    initialRawParams = initialRawParams.filter(function tmp(obj) { return !globalRemovedParams.includes(obj); });
+    // Adding parameters
+    initialRawParams = initialRawParams.concat(globalAddedParams);
+    return window.location.pathname + '?' + initialRawParams.join('&');
+  }
+
+  function applyMultiFacetsSelection() {
+    defaultContext.getElementsByClassName('applyMultiFacetsSelection')[0]
+      .removeEventListener('click', applyMultiFacetsSelection);
+    callbackOnApply();
+    window.location.assign(getHrefWithNewParams());
+  }
+
+  function dateSelectorInit() {
+    document.querySelectorAll('div.facet form .date-fields').forEach((elem) => {
+      dateSelectorIds.push(elem.parentElement.id);
+      elem.parentElement.addEventListener('submit', function switchAction(e) {
+        if (isMultiFacetsSelectionActivated) {
+          e.preventDefault();
+        }
+      });
+    });
+  }
+
+  function facetSelectionStyling(elem) {
+    if (elem.classList.contains('exclude')) {
+      elem.classList.toggle('selected');
+    } else {
+      if (elem.classList.contains('facet')) {
+        elem.classList.toggle('active');
+      } else if (elem.parentElement.classList.contains('facet')) {
+        elem.parentElement.classList.toggle('active');
+      }
+
+      let icon = elem.querySelector('.icon');
+      if (icon !== null && icon.classList.contains('fa-check-square-o')) {
+        icon.classList.remove('fa-check-square-o');
```

Comment:
Do you prefer the new way?

Reply:
I think this is progress, but it's still only going to work with font-based icons, and VuFind supports other methods.

To demonstrate:

1.) Set the theme to bootprint3
2.) Edit themes/bootprint3/theme.config.php and in the icon configuration, add `'facet-checked' => 'Fugue:tick.png'`. You'll see that when you apply orFacets in this configuration, they get green checkmarks instead of the checkbox, but if you try to apply multi-select, the icons don't work.

This is why I thought it might make sense to use a data element on the icon container to indicate the checked/unchecked state (so you don't have to do any complex analysis) and then just replace the innerHTML with the pre-rendered icons when the state is toggled. I think this would require the JS code to be less aware of the actual contents of the markup, which may make it less likely to break in future if any implementation details change, or if people have significantly customized their icon displays.

Reply:
No, that's not what I'm proposing. When you call `VuFind.icon()`, it gives you the fully rendered icons for "checked" and "unchecked." I propose that if we include a "data-checked" property on the icons in the HTML, then we know which state the icon represents. The Javascript could then use the data element to identify which state is being changed, and use innerHTML to replace the icon with the opposite icon. The only increase in the page output would be the `data-checked="true|false"` on each facet value.